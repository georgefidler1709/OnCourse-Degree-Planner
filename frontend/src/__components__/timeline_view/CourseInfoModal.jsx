import React from "react";
import Modal from "react-bootstrap/Modal";
import Button from "react-bootstrap/Button";

function CourseInfoModal(props) {
    const year = '2020'
    const handbook = `https://www.handbook.unsw.edu.au/undergraduate/courses/${year}`
    return (
      <Modal
        {...props}
        size="lg"
        aria-labelledby="contained-modal-title-vcenter"
        centered
      >
        <Modal.Header closeButton>
          <Modal.Title id="contained-modal-title-vcenter">
            {props.courseId}
          </Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <a href={`${handbook}/${props.courseId}`}>HERE</a>
        </Modal.Body>
        <Modal.Footer>
          <Button onClick={props.onHide}>Close</Button>
        </Modal.Footer>
      </Modal>
    );
  }

  export default CourseInfoModal