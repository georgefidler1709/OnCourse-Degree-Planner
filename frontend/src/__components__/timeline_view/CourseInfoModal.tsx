import React from "react";
import Modal from "react-bootstrap/Modal";
import Button from "react-bootstrap/Button";

const year = '2020'
const handbook = `https://www.handbook.unsw.edu.au/undergraduate/courses/${year}`


interface CourseInfoModalProps {
  index: number;
  onHide: () => void;
  removeCourse: (s: string) => void; 
  show: boolean;
  code: string;
  name: string;
  prereqs: string;
  coreqs: string;
  equivalents: string;
  exclusions: string;
}

function displayCourseReqs(reqs: string, req_type: string) {

  const noBullet = {
    "list-style-type" : "none",
  } as React.CSSProperties;
  
  return (
    <div id={req_type}>
      <h5>{req_type + ":"}</h5>
      {reqs ? (
        <ul>
          {reqs.split("\n").map(req => <li key={req}>{addLinks(req)}</li>)}
        </ul>
      ) : (
        <ul style={noBullet}>
          <li>None</li>
        </ul>
      )} 
    </div>
  )
}

function addLinks(req: string) {
  req = req.replace(/[()]/g, '');
  return req.split(' ').map(word => {
    if(word === "OR" || word === "AND" || word === "") return " " + word + " "
    return (<a key={word} href={`${handbook}/${word}`}>{word}</a>)
  })
}

function CourseInfoModal(props: CourseInfoModalProps) {
    return (
      <Modal
        {...props}
        size="lg"
        aria-labelledby="contained-modal-title-vcenter"
        centered
      >
        <Modal.Header closeButton>
          <Modal.Title id="contained-modal-title-vcenter">
            {props.code + " - "}
            {props.name}
          </Modal.Title>
        </Modal.Header>
        <Modal.Body>
          {displayCourseReqs(props.prereqs, "Prereqs")}
          {displayCourseReqs(props.coreqs, "Coreqs")}
          {displayCourseReqs(props.equivalents, "Equivalents")}
          {displayCourseReqs(props.exclusions, "Exclusions")}
          <hr/>
          <a href={`${handbook}/${props.code}`}>More Info</a>
        </Modal.Body>
        <Modal.Footer>
          <Button onClick={props.onHide}>Close</Button>
          <Button variant="danger" onClick={() => {
            props.onHide()
            props.removeCourse(props.code)
          }}>Remove</Button>
        </Modal.Footer>
      </Modal>
    );
  }

  export default CourseInfoModal
