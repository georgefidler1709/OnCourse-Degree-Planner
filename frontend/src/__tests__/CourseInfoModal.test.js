import React from 'react'
import {  mount } from 'enzyme';
import CourseInfoModal from '../__components__/timeline_view/CourseInfoModal';
import Button from "react-bootstrap/Button";

const mockCourse = {
  key: "key",
  index: 0,
  code: "COMP3121",
  coreqs: "",
  equivalents: "COMP3821, COMP9801, COMP3120, COMP9101",
  exclusions: "",
  name: "Algorithms and Programming Techniques",
  prereqs: "(COMP1927 OR COMP2521)",
}


console.error = jest.fn();

let modalShow = true;
let setModalShow = jest.fn()
describe('Rendering a course info popup', () => {
  it('renders correctly', () => {
    const wrapper = mount(<CourseInfoModal
      show={modalShow}
      onHide={() => setModalShow(false)}
      {...mockCourse}
    />);
    expect(wrapper).toMatchSnapshot();

    wrapper.unmount();
  });

  it('will be removed when close button clicked', () => {
      const wrapper = mount(<CourseInfoModal
        show={modalShow}
        onHide={() => setModalShow(false)}
        {...mockCourse}
      />);
      
      wrapper.find(Button).first().simulate('click');
      expect(setModalShow).toHaveBeenCalledWith(false);

      wrapper.unmount();
    });

    it('display reqs correctly', () => {
      const wrapper = mount(<CourseInfoModal
        show={modalShow}
        onHide={() => setModalShow(false)}
        {...mockCourse}
      />);
      //enzyme does not include html to the line break between : and COMP1927 is preserved
      expect(wrapper.find("#Prereqs").text()).toBe('Prereqs:COMP1927 OR COMP2521')
      wrapper.unmount();
    });
});