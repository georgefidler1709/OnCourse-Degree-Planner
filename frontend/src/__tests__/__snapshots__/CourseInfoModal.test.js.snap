// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering a course info popup renders correctly 1`] = `
<Bootstrap(Modal)
  aria-labelledby="contained-modal-title-vcenter"
  centered={true}
  code="COMP3121"
  coreqs=""
  equivalents="COMP3821, COMP9801, COMP3120, COMP9101"
  exclusions=""
  index={0}
  name="Algorithms and Programming Techniques"
  onHide={[Function]}
  prereqs="(COMP1927 OR COMP2521)"
  show={true}
  size="lg"
>
  <ModalHeader
    closeButton={true}
    closeLabel="Close"
  >
    <ModalTitle
      id="contained-modal-title-vcenter"
    >
      COMP3121 - 
      Algorithms and Programming Techniques
    </ModalTitle>
  </ModalHeader>
  <ModalBody>
    <div
      id="Prereqs"
    >
      <h5>
        Prereqs:
      </h5>
      <ul>
        <li
          key="(COMP1927 OR COMP2521)"
        >
          <a
            href="https://www.handbook.unsw.edu.au/undergraduate/courses/2020/COMP1927"
            key="COMP1927"
          >
            COMP1927
          </a>
           OR 
          <a
            href="https://www.handbook.unsw.edu.au/undergraduate/courses/2020/COMP2521"
            key="COMP2521"
          >
            COMP2521
          </a>
        </li>
      </ul>
    </div>
    <div
      id="Coreqs"
    >
      <h5>
        Coreqs:
      </h5>
      <ul
        style={
          Object {
            "list-style-type": "none",
          }
        }
      >
        <li>
          None
        </li>
      </ul>
    </div>
    <div
      id="Equivalents"
    >
      <h5>
        Equivalents:
      </h5>
      <ul>
        <li
          key="COMP3821, COMP9801, COMP3120, COMP9101"
        >
          <a
            href="https://www.handbook.unsw.edu.au/undergraduate/courses/2020/COMP3821,"
            key="COMP3821,"
          >
            COMP3821,
          </a>
          <a
            href="https://www.handbook.unsw.edu.au/undergraduate/courses/2020/COMP9801,"
            key="COMP9801,"
          >
            COMP9801,
          </a>
          <a
            href="https://www.handbook.unsw.edu.au/undergraduate/courses/2020/COMP3120,"
            key="COMP3120,"
          >
            COMP3120,
          </a>
          <a
            href="https://www.handbook.unsw.edu.au/undergraduate/courses/2020/COMP9101"
            key="COMP9101"
          >
            COMP9101
          </a>
        </li>
      </ul>
    </div>
    <div
      id="Exclusions"
    >
      <h5>
        Exclusions:
      </h5>
      <ul
        style={
          Object {
            "list-style-type": "none",
          }
        }
      >
        <li>
          None
        </li>
      </ul>
    </div>
    <hr />
    <a
      href="https://www.handbook.unsw.edu.au/undergraduate/courses/2020/COMP3121"
    >
      More Info
    </a>
  </ModalBody>
  <ModalFooter>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      type="button"
      variant="primary"
    >
      Close
    </Button>
    <Button
      active={false}
      disabled={false}
      onClick={[Function]}
      type="button"
      variant="danger"
    >
      Remove
    </Button>
  </ModalFooter>
</Bootstrap(Modal)>
`;
